### PR TITLE
Add Pseudo Least Recently Used Cache replacement policy

### PR DESCRIPTION
**Pull request by brsf11**
Add PLRU replacement policy, which is more widely used in real-world hardware than ideal LRU.

image

### CODE REVIEW

File: src/machine/memory/cache/cache_policy.cpp [hunk 1]
```diff
@@ -128,4 +132,37 @@ size_t CachePolicyRAND::select_way_to_evict(size_t row) const {
...
+        row.resize((2 << associativityCLog2)-1,0); // Initially point to block 0
+    }
+}
+
+void CachePolicyPLRU::update_stats(size_t way, size_t row, bool is_valid) {
+    UNUSED(is_valid)
+    uint32_t plru_idx = 0;
+    auto &row_ptr = plru_ptr.at(row);
+    for (uint32_t i = 0; i < associativityCLog2; i++) {
+        row_ptr[plru_idx] = ((way >> (associativityCLog2 - 1 - i)) & 1) ? 0 : 1;
```
Comment (jdupak) [inline src/machine/memory/cache/cache_policy.cpp:149]: Could you add some comments here? I think it is correct, but it is rather hard to read. Maybe a little common subexpression extraction to named variables would help.

**Review by jdupak — CHANGES_REQUESTED** [review]
Thank you for your contribution, it looks very good overall! There are just some small issues.
Also it would be very good to extend and regenerate the tests is https://github.com/cvut/qtrvsim/blob/master/src/machine/memory/cache/cache.test.cpp

Also please make sure, that the code is formatted according to provided .clang-format. Thank you

File: src/machine/memory/cache/cache_policy.cpp (hunk 1 above)
Comment (brsf11) [reply src/machine/memory/cache/cache_policy.cpp:149]: Sure no problem

File: src/machine/memory/cache/cache_policy.cpp [hunk 2]
```diff
@@ -128,4 +132,37 @@ size_t CachePolicyRAND::select_way_to_evict(size_t row) const {
     UNUSED(row)
     return std::rand() % associativity; // NOLINT(cert-msc50-cpp)
 }
+
+CachePolicyPLRU::CachePolicyPLRU(size_t associativity, size_t set_count)
+    : associativity(associativity), associativityCLog2(std::ceil(log2((float)associativity))) {
+    plru_ptr.resize(set_count);
+    for (auto &row : plru_ptr) {
+        row.resize((2 << associativityCLog2)-1,0); // Initially point to block 0
```
[the remark this answers is not in the document]
Comment (brsf11) [reply src/machine/memory/cache/cache_policy.cpp:140]: Yes, it's a typo. Thanks for pointing out!

**Comment by brsf11** [comment]
I've pushed the commit for the typo, comment and cache test. For the clang-format, I'm not sure if my tool works properly since I've seen some modification on previous code when I used the format tool.

e.g.

```c++
switch (config->replacement_policy()) {
case CacheConfig::RP_RAND:
    return std::make_unique<CachePolicyRAND>(config->associativity());
case CacheConfig::RP_LRU:
    return std::make_unique<CachePolicyLRU>(
        config->associativity(), config->set_count());
case CacheConfig::RP_LFU:
    return std::make_unique<CachePolicyLFU>(
        config->associativity(), config->set_count());
case CacheConfig::RP_PLRU:
    return std::make_unique<CachePolicyPLRU>(
        config->associativity(), config->set_count());
}
```
will be converted to
```c++
switch (config->replacement_policy()) {
case CacheConfig::RP_RAND:
    return std::make_unique<CachePolicyRAND>(config->associativity());
case CacheConfig::RP_LRU:
    return std::make_unique<CachePolicyLRU>(config->associativity(), config->set_count());
case CacheConfig::RP_LFU:
    return std::make_unique<CachePolicyLFU>(config->associativity(), config->set_count());
case CacheConfig::RP_PLRU:
    return std::make_unique<CachePolicyPLRU>(config->associativity(), config->set_count());
}
```

**Comment by jdupak** [comment]
> I've pushed the commit for the typo, comment and cache test. For the clang-format, I'm not sure if my tool works properly since I've seen some modification on previous code when I used the format tool.
> 
> e.g.
> 
> ```c++
> switch (config->replacement_policy()) {
> case CacheConfig::RP_RAND:
>     return std::make_unique<CachePolicyRAND>(config->associativity());
> case CacheConfig::RP_LRU:
>     return std::make_unique<CachePolicyLRU>(
>         config->associativity(), config->set_count());
> case CacheConfig::RP_LFU:
>     return std::make_unique<CachePolicyLFU>(
>         config->associativity(), config->set_count());
> case CacheConfig::RP_PLRU:
>     return std::make_unique<CachePolicyPLRU>(
>         config->associativity(), config->set_count());
> }
> ```
> 
> will be converted to
> 
> ```c++
> switch (config->replacement_policy()) {
> case CacheConfig::RP_RAND:
>     return std::make_unique<CachePolicyRAND>(config->associativity());
> case CacheConfig::RP_LRU:
>     return std::make_unique<CachePolicyLRU>(config->associativity(), config->set_count());
> case CacheConfig::RP_LFU:
>     return std::make_unique<CachePolicyLFU>(config->associativity(), config->set_count());
> case CacheConfig::RP_PLRU:
>     return std::make_unique<CachePolicyPLRU>(config->associativity(), config->set_count());
> }
> ```

Yeah, for some parts may not be correctly formatted now. That is ok.

**Review by jdupak — APPROVED** [review]
I did some commit cleanup and proper reformatting myself. Otherwise very good work, thank you!

**Pull request merged**